dolt/diffs.py

```diff
@@ -39,79 +39,103 @@ def two_dot_diffs(from_commit=None, to_commit=None):
+                # Annotate each row with a JSON-ified diff
+                diff=RawSQL(  # nosec
+                    f"""SELECT JSON_OBJECT("root", "from", {json_diff_fields(tbl_name)})
+                        FROM dolt_commit_diff_{tbl_name}
+                        WHERE to_commit = %s AND from_commit = %s
+                        AND from_id = {tbl_name}.id """,
+                    (to_commit, from_commit),
                     output_field=models.JSONField(),
-                ),
+                )
             )
             # "time-travel" query the database at `from_commit`
             .using(db_for_commit(from_commit))
         )
-
         diff_rows = sorted(list(to_queryset) + list(from_queryset), key=lambda d: d.pk)
         if len(diff_rows) == 0:
             continue
 
         diff_view_table = DiffListViewFactory(content_type).get_table_model()
         diff_results.append(
             {
-                "name": f"{factory.source_model_verbose_name} Diffs",
+                "name": f"{verbose_name} Diffs",
                 "table": diff_view_table(diff_rows),
-                "added": diffs.filter(diff_type="added").count(),
-                "modified": diffs.filter(diff_type="modified").count(),
-                "removed": diffs.filter(diff_type="removed").count(),
+                **diff_summary_for_table(tbl_name, from_commit, to_commit),
             }
         )
     return diff_results
 
 
-def diff_annotation_query_fields(model):
-    """diff_annotation_query_fields returns all of the column names for a model and turns them into to_ and from_ fields."""
-    names = [
-        f.name
-        for f in model._meta.get_fields()
-        # field names containing "__" cannot be
-        # diffed as Django interprets kwargs with
-        # "__" as lookups
-        if "__" not in f.name
-    ]
-    return {name: F(name) for name in names}
+def diff_summary_for_table(table, from_commit, to_commit):
+    """diff_summary_for_table returns the diff summary for table, for the commits from_commit and to_commit."""
+    summary = {
+        "added": 0,
+        "modified": 0,
+        "removed": 0,
+    }
+    with connection.cursor() as cursor:
+        cursor.execute(
+            f"""SELECT diff_type, count(diff_type) FROM dolt_commit_diff_{table}  # nosec
+                WHERE to_commit = %s AND from_commit = %s
+                GROUP BY diff_type ORDER BY diff_type""",  # nosec
+            (to_commit, from_commit),
+        )
+        for pair in cursor.fetchall():
+            summary[pair[0]] = pair[1]
+    return summary
+
+
+def json_diff_fields(tbl_name):
+    """
+    json_diff_fields returns all of the column names for a model
+    and turns them into to_ and from_ fields.
+    """
+    with connection.cursor() as cursor:
+        cursor.execute(f"DESCRIBE dolt_commit_diff_{tbl_name}")
+        cols = cursor.fetchall()
+    pairs = [f"'{c[0]}', dolt_commit_diff_{tbl_name}.{c[0]}" for c in cols]
```

Comment:
Just in case this gets called often or there are a lot of rows, let's make this a generator expression instead of a list comprehension so a new list isn't created in place only to be discarded in the next line.

```suggestion
    pairs = (f"'{c[0]}', dolt_commit_diff_{tbl_name}.{c[0]}" for c in cols)
```